src/controller/CHIPDeviceControllerFactory.cpp

```diff
@@ -487,6 +491,18 @@ void DeviceControllerSystemState::Shutdown()
         mSessionMgr = nullptr;
     }
 
+    if (mReportScheduler != nullptr)
+    {
+        chip::Platform::Delete(mReportScheduler);
```

Comment:
Hmm.  There is no shutdown on the scheduler, so I assume its destructor is expected to clean up?  I don't see `~SynchronizedReportSchedulerImpl` canceling its timer.

Reply:
It does call the parent class destroyer at the end, which will call UnregisterAllHandlers, which calls OnReadHandlerDestroyed, which will call CancelReport, which will call CancelTimer

Reply:
No, that does not work.

First the subclass destructor runs.  It does nothing.  The vtable pointer is then pointed to the superclass vtable, since the subclass has been destroyed (!).  This is also why calling virtual methods in destructors is a bit suspect....

Then the superclass destructor runs.  It calls `UnregisterAllHandlers`, which as you note calls `OnReadHandlerDestroyed` for each handler.  The vtable is now for the superclass, so we end up in `ReportSchedulerImpl::OnReadHandlerDestroyed`.  This calls `ReportSchedulerImpl::CancelReport` (`CancelReport` is not virtual anyway), which tries to cancel the timer with the node as the context.  But the context should have been the `SynchronizedReportSchedulerImpl`.  So that one timer never gets canceled.